spec/p2p/reactor/README.md

```diff
@@ -0,0 +1,224 @@
+`Switch.AddReactor(name string, reactor Reactor)` method, where `name` can be
+an arbitrary string.
+
+The registration should happen before the node, in general, and the p2p layer,
+in particular, are started.
+In other words, there is no support for registering a reactor on a running node:
+reactors must be registered as part of the setup of a node.
+
+```abnf
+registration    = get-channels set-switch
+```
+
+The p2p layer retrieves from the reactor a list of channels the reactor is
+responsible for, using the `GetChannels()` method.
+The reactor implementation should thereafter expect the delivery of every
+message received by the p2p layer in the informed channels.
+
+The second method `SetSwitch(Switch)` concludes the handshake between the
+reactor and the p2p layer.
+The `Switch` is the main component of the p2p layer, being responsible for
+establishing connections with peers and routing messages.
+
+## Service interface
+
+A reactor should implement the [`Service`](../../../libs/service/service.go) interface,
+in particular, a startup `OnStart()` and a shutdown `OnStop()` methods:
+
+```abnf
+start           = registration on-start *peer-management on-stop
+```
+
+As part of the startup of a node, all registered reactors are started by the p2p layer.
+And when the node is shutdown, all registered reactors are stopped by the p2p layer.
+Observe that the `Service` interface specification establishes that a service
+can be started and stopped only once.
+So before being started or once stopped by the p2p layer, the reactor should
+not expect any interaction.
+
+## Peer management
+
+The core of a reactor's operation is the interaction with peers or, more
+precisely, with companion reactors operating in peers connected to the node.
+The grammar extract below represents the interaction of the reactor with a
+single peer:
+
+```abnf
+; Refers to a single peer, a reactor should support multiple concurrent peers
+peer-management = init-peer start-peer stop-peer
+```
+
+The p2p layer informs all registered reactors when it establishes a connection
+with a `Peer`, using the `InitPeer(Peer)` method.
+When this method is invoked, the `Peer` has not yet been started, namely the
+routines for sending messages to and receiving messages from the peer are not running.
+This method should be used to initialize state or data related to the new
+peer, but not to interact with it.
+
+The next step is to start the communication routines with the new `Peer`.
+As detailed in the following, this procedure may or may not succeed.
+In any case, the peer is eventually stopped, which concludes the management of
+that `Peer` instance.
+
+## Start peer
+
+Once a `Peer` is initialized by every reactor, the p2p layer starts the peer's
+communication routines and adds the `Peer` to the set of connected peers.
+If both steps are concluded without errors, the reactor's `AddPeer(Peer)` is invoked:
+
+```abnf
+start-peer      = [*receive] (connected-peer / start-error)
```

Comment:
For the minor comment: yes, this is redundant. In a previous version we had `[receive]` with a comment that this could actually happen more than once. But then we realize that we cannot provide an "upper bound" on the number of times this happens.

I would maintain this wording which, although redundant, highlights that receiving messages here is not the "expected" behavior, but a possible behavior that has to be taken into account.